### PR TITLE
Soc `PROCAR` handling and more

### PR DESCRIPTION
**Pull request by kavanase**
This is mainly a fix to update the `PROCAR` handling in `easyunfold` (also updating to make the `Procar` object JSON serializable), and in particular updating to handle SOC (ncl) `PROCAR`s. This is now dynamically determined within the `Procar` parsing code, so when plotting via the CLI the user doesn't need to specify `--ncl` (previously this information wasn't passed on to the `PROCAR` parsing anyway).

This fixes the cases I found where plotting failed due to crashes when trying to parse SOC `PROCAR`s, now fixed, giving some nice plots like this:
image

I've also checked through with previous example cases (NaBiS$_2$, MgO) to confirm the code is working as expected.

This also implements some other fixes:
- Removing all x/y ticks in the `sumo` DOS plots.
- Using the `sumo` orbital colour settings if atom projections aren't used and the user doesn't specify
- Docs updates
- Deprecation notices for `vscale`

### CODE REVIEW

**Comment by kavanase** [comment]
This was based off the recently merged `dos-plotting` branch, so the new commits are the ones from [f821de2](https://github.com/SMTG-UCL/easyunfold/pull/28/commits/f821de2178d58fe3fed5ae0b22886bbd7cc59023) on

**Comment by kavanase** [comment]
Hi @zhubonan!
So when plotting the atom projections for my Cs$_2$(Sn/Ti)X$_6$ alloys, I realised something a little odd. If you look at the `Sn s` - `Br -p` conduction band in this GIF:
CSTB_Easyunfold
or static plot below, you can see that intensity of this band appears quite weak (and doesn't look like a mix between green/blue, just faded grey), and also in the pure Cs$_2$SnBr$_6$ material, it looks basically just blue, despite it being a 50/50 mix of `Sn s` (green) and `Br p` (blue) as you can see from the DOS:
Sn50_Ti50_unfold_projections_intensity_3

But if I swap the colour choice, so that `Ti` is green and `Sn` is red, then the colour looks more reasonable/intuitive and more intense/visible:
Sn50_Ti50_unfold_projections_intensity_3_swapped_colours

I dug around for a long while trying to find the origin of this, and how it could be improved. Basically it's related to how the interpolation is done in the transformed LAB space, which can result in some non-intuitive behaviour like this, depending on the colours being mixed (as you end up interpolating into the 'white' origin centre in LAB). I checked and this is similarly the case in `sumo` (which uses the same interpolation strategy), though the lack of additional transparency (due to the spectral function intensity) and underlying black line makes it much less severe in their case (though still not totally intuitive that 50/50 green/blue gives you that colour I think):
image

Colorspace  |  Red-Green-Blue  |  Pastel
:-------------------------:|:-------------------------:|:-------------------------:
LAB | unfold_projections_intensity_3_SK_dark_lab | unfold_projections_intensity_3_SK_dark_other_lab
HSV | unfold_projections_intensity_3_SK_dark_hsv | unfold_projections_intensity_3_SK_dark_other_hsv
RGB | unfold_projections_intensity_3_SK_dark_rgb | unfold_projections_intensity_3_SK_dark_other_rgb
XYZ | unfold_projections_intensity_3_SK_dark_xyz| unfold_projections_intensity_3_SK_dark_other_xyz
LABlch | unfold_projections_intensity_3_SK_dark_lablch  |  unfold_projections_intensity_3_SK_dark_other_lablch
LUVlch | unfold_projections_intensity_3_SK_dark_luvlch | unfold_projections_intensity_3_SK_dark_other_luvlch

So I did a systematic check through the different possible colourspaces, using either red-green-blue or pastel red-green-blue as the basis, with these weight choices (covering the typical mixing spectrum)
```
    weights = np.array([
        [1.0, 0.0, 0.0],
        [1.0, 0.5, 0.0],
        [0.5, 0.5, 0.0],
        [1.0, 1.0, 0.0],
        [0.5, 1.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.0, 1.0, 0.5],
        [0.0, 0.5, 0.5],
        [0.0, 1.0, 1.0],
        [0.0, 0.5, 1.0],
        [0.0, 0.0, 1.0],
        [0.5, 0.0, 1.0],
        [0.5, 0.0, 0.5],
        [1.0, 0.0, 1.0],
        [1.0, 0.0, 0.5],
        [0.5, 0.5, 0.5]])
```
Using the red-green-blue basis:
RGB_Colours

Using the pastel basis:
Other_colours

My conclusions/opinions would be that (1) pastel colours are nicer, (2) RGB/XYZ/LAB make intuitive sense mostly, so suggest sticking with LAB as the default colourspace, using the pastel colours if the number of projected species is <= 3 (else red, green, blue, purple, orange, yellow as suggested by colorbrewer.org) and (3) making colorspace an optional parameter at the CLI level.
I've added these updates to the code here. If you're happy with all this then I'll update the atom/orbital-projected plots in [MgO](https://smtg-ucl.github.io/easyunfold/examples/example_mgo.html) and [NaBiS2](https://smtg-ucl.github.io/easyunfold/examples/example_nabis2.html#atom-projected-unfolded-band-structure-with-dos) examples, and we can get this merged! 😃

**Comment by kavanase** [comment]
Jupyter notebook for testing and plotting these colours here if useful:
[easyunfold_colourmap_testing.ipynb.zip](https://github.com/SMTG-UCL/easyunfold/files/12279535/easyunfold_colourmap_testing.ipynb.zip)

**Comment by kavanase** [comment]
Ok great! I'll update the docs to match the new defaults, and add the 'example outputs' mini section discussed offline, and then this is ready to be merged ✌️ 🏎️

**Comment by kavanase** [comment]
All final bits discussed now implemented @zhubonan, this is ready to be merged! 🚀

**Comment by kavanase** [comment]
Some last formatting changes added there @zhubonan.
The example outputs mini-section can be seen here: https://github.com/SMTG-UCL/easyunfold/tree/SOC-procar-handling
All ready to go.

File: easyunfold/procar.py [hunk 1]
```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
...
-                line = re.sub(r'(\d)-', r'\1 -', line)
-                tokens = line.strip().split()
-                kvecs.append([float(val) for val in tokens[-6:-3]])
-                kweights.append(float(tokens[-1]))
-        self.occs = np.array(occs)
-        self.kvecs = np.array(kvecs)
-        self.kweights = np.array(kweights)
-        self.eigenvalues = np.array(energies)
+
+        with fobj:
```
Comment (zhubonan) [inline easyunfold/procar.py:63]: We should not preemptively close the file object here, e.g. the stream may be used by other  functions later. 
Closing the file object can be moved to the `read` method instead, if the file object is opened by that method.

File: easyunfold/procar.py [hunk 2]
```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
...
+                self.proj_names = line.strip().split()[1:-1]
+                break
+
+    def read(self, fobjs_or_paths):
+        """Read and amalgamate the data from a list of PROCARs"""
+
+        def open_file(fobj_or_path):
+            if isinstance(fobj_or_path, (str, Path)):
+                return open(fobj_or_path, encoding='utf-8')  # closed later
+            return fobj_or_path  # already a file-like object, just return it
```
Comment (zhubonan) [inline easyunfold/procar.py:164]:
```suggestion
            if isinstance(fobj_or_path, (str, Path)):
                return open(fobj_or_path, encoding='utf-8'), True # closed later
            return fobj_or_path, False  # already a file-like object, just return it
```
Can record if the file object is opened by the function or not.

File: easyunfold/procar.py [hunk 3]
```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
...
+
+        parsed_kpoints = None
+        occs_list, kvecs_list, kweights_list = [], [], []
+        eigenvalues_list, proj_data_list, proj_xyz_list = [], [], []
+        for i, fobj_or_path in enumerate(fobjs_or_paths):
+            # Note: If PROCAR parsing becomes a significant bottleneck for people (e.g. with several HSE06+SOC PROCARs),
+            # this could be parallelized (somewhat) with multiprocessing. The actual file parsing in _read() is currently
+            # serial so no easy wins there, but could at least parallelise over the list of PROCARs
+            fobj = open_file(fobj_or_path)
```
Comment (zhubonan) [inline easyunfold/procar.py:173]:
```suggestion
            fobj, opened = open_file(fobj_or_path)
```

File: easyunfold/procar.py [hunk 4]
```diff
@@ -32,85 +38,198 @@ def __init__(self, fobj_or_path=None, is_soc=False):
...
+                self._read_header_nion_proj_names(fobj)
+
+            current_nspins = self.nspins  # check spin consistency between PROCARs
+            nspins, occs, kvecs, kweights, eigenvalues, proj_data, proj_xyz, parsed_kpoints = self._read(fobj,
+                                                                                                         parsed_kpoints=parsed_kpoints)
+            if current_nspins is not None and current_nspins != nspins:
+                raise ValueError(f'Mismatch in number of spins in PROCARs supplied: ({nspins} vs {current_nspins})!')
+
+            if isinstance(fobj_or_path, (str, Path)):
+                fobj.close()  # if file was opened in this loop, close it
```
Comment (zhubonan) [inline easyunfold/procar.py:184]:
```suggestion
            if opened:
                fobj.close()  # if file was opened in this loop, close it
```
Something like this.

Comment (zhubonan) [reply easyunfold/procar.py:184]: I think the current code is fine - just that there is no need to use `with` inside `_read` method.

**Review by zhubonan — CHANGES_REQUESTED** [review]
Thanks for the improvements!

Just one minor comment regarding closing file object in `_read` - I think `with` clause if not needed. If `Procar` receives file objects (can be streams) rather than file names/paths, we should keep these objects open .

**Comment by zhubonan** [comment]
Also, now `Procar` is `MSONable`, we should be able to store it inside the `easyunfold.json` file and/or as a separate file? This way people can get rid of the PROCAR, but I am not sure the space-saving would be much if everything is compressed.

**Comment by kavanase** [comment]
Hi @zhubonan!
Have now removed the `with` statement from the `_read` method as requested.

About saving the `PROCAR` info to JSON, I was thinking this as well and tried it, but it turned out from testing a few cases that the IO speed from saving/loading the JSON was basically the same as reading the PROCAR from scratch, so there was no benefit in speedup from saving the PROCAR info to JSON. Also the file size was essentially the same as the PROCAR file size, so no file size reduction benefit either.
A disadvantage to this as well is that if it was saved in the `easyunfold.json` file then it would cause normal plotting functions (without projections) to be much slower as it would take much longer to load the large JSON file.

That said, something related I was going to mention, I think the loops in `PROCAR` parsing and `spectral_function_from_weight_sets` could be significantly sped up by only parsing bands with energies within the `emin` - `emax` range. Currently they parse all bands and then subselect from there. 
I noticed this with the Cs$_2$Sn/TiX$_6$ alloys when trying to debug the colour projections, where I was looking at the `eng` and `sf_rgb` arrays directly, and realised they stretched from -37 to +8 eV, even though only plotting from -5 to +5 eV. This is typical of most VASP calculations, so might be worth keeping in mind as a potential route to speeding up parsing for heavier cases (like atom-projected SOC supercells). Can def wait till after the JOSS submission though! 😅

**Review by zhubonan — APPROVED** [review]
All good!

**Comment by zhubonan** [comment]
>  I think the loops in PROCAR parsing and spectral_function_from_weight_sets could be significantly sped up by only parsing bands with energies within the emin - emax range. Currently they parse all bands and then subselect from there.

Good point 👍 , can probably implement it later!

Yeah , I agree there is not much point to save PROCAR in JSON for now.

**Comment by zhubonan** [comment]
Merged! Thanks again for fixing the PROCAR SOC bug and adding secondary axis DOS plot 🎊

**Pull request merged**